---
title: "PLATCOV Statistical Analysis"
author: "James Watson"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document: 
    toc: yes
    fig_caption: yes
    keep_md: yes
    number_sections: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache = F, cache.comments = FALSE,
                      echo = F, include = TRUE, 
                      fig.width = 8, fig.height = 8,
                      fig.pos = 'H',dev = 'png', dpi = 300)

library(rstan)
library(brms)
library(loo)
library(RColorBrewer)
library(reshape2)
library(plotrix)
library(lme4cens)
library(censReg)
```

TODO list:

- Load randomization data files and cross check
- Make generic function for data summaries

# Preamble

This Statistical Analysis Plan (SAP) is written for the PLATCOV trial (registered at ClinicalTrials.gov: <https://clinicaltrials.gov/ct2/show/NCT05041907>).

Data preparation is done in a different R script called data_prep.R. This Markdown script assumes that the data are saved in a .csv file *interim_dat.csv* in long format. The file *interim_dat.csv* contains the patient clinical and viral load data with the following column headers:

-   ID: anonymized patient id code
-   Time: time from randomization
-   Trt: treatment allocation as written in CRF
-   Site: site at enrollment
-   Timepoint_ID: Day of study (integer 0 to 14)
-   BARCODE: unique sample id code (in the following format: PLT-site-number-swab type-timepoint)
-   Swab_ID: RTS or TSL (right versus left tonsil)
-   Plate: unique Plate ID for the PCR assay (matching with plate identifiers in interim_control_dat.csv)
-   Rand_date: date of randomization
-   Any_dose: (0/1) any doses of any manufacturer received
-   N_dose: integer number of doses received (any manufacturer)
-   Antibody_test: 0/1 (negative/positive for SARS-CoV-2 antibody rapid test)
-   Weight (kg)
-   BMI: kg/weight\^2
-   Age: (years - has to be between 18-50)
-   Sex: 0/1 (male: 1; female/other: 0)
-   Symptom_onset: time since onset of symptoms (days)
-   Variant: variant of concern (using standard WHO terminology for the main lineages, reference will be the predominant variant in the data set at the start of the study)
-   CT_NS: observed CT value for the N/S gene
-   CT_RNaseP: observed CT value for the human RNase P gene
-   Per_protocol_sample: whether at the time of sampling the patient was still in per protocol with respect to drug dosing
-   IgG: + IgG band on the LFT
-   IgM: + IgM band on the LFT
-   log10_viral_load: log10 number of viral copies per mL (estimated from control samples using a mixed effects model)
-   log10_cens_vl: censoring value

## PCR QC

```{r}
source('PCR_quality_checks.R')
```


## Computational setup

```{r}
## information on software/hardware used
version
sessionInfo()

study_threshold = 1.125

rstan_options(auto_write = TRUE)
## parameters for the analysis
Dmax = 8
RUN_MODELS = F

source('functions.R')
```

# Overview of data
## Load data

```{r}
screen_dat = read.csv('interim_screening_dat.csv')
platcov_dat = read.csv('interim_dat.csv')

# remove the saliva samples
platcov_dat = subset(platcov_dat, Swab_ID != 'Saliva')

# make per protocol summary for all patients
PP=merge(aggregate(Timepoint_ID ~ ID+Trt,
                   platcov_dat[platcov_dat$Per_protocol_sample==1, ], max),
         aggregate(Per_protocol_sample ~ ID, platcov_dat, sum),by = 'ID')
PP$Include_mITT = PP$Timepoint_ID>=2


# We remove patients who only have undetectable virus
xx_undetectble = table(platcov_dat$ID, platcov_dat$CT_NS==40)
ids_neg = names(which(xx_undetectble[,1]==0))
writeLines(sprintf('All negative samples for id: %s', ids_neg))

# Exclude from mITT pop
PP$Include_mITT[PP$ID %in% ids_neg] = F

platcov_dat = platcov_dat[!platcov_dat$ID %in% ids_neg, ]

writeLines(sprintf('PCR database contains data from %s patients',nrow(PP)))
```





## Data summaries

Time from symptom onset

```{r symptom_onset}
bvl=aggregate(log10_viral_load ~ ID + Symptom_onset + Site, 
              data = platcov_dat[platcov_dat$Timepoint_ID==0,],FUN = mean)
par(mfrow=c(1,2),las=1)
for(ss in c('th001','br003')){
  ind = bvl$Site==ss
  boxplot(bvl$log10_viral_load[ind] ~ round(bvl$Symptom_onset[ind]),main=ss,
          ylim = c(3,8), xlab = 'Days from symptom onset', 
          ylab = 'Baseline viral load (log10)')
  print(summary(lm(bvl$log10_viral_load[ind] ~ bvl$Symptom_onset[ind])))
}
```


Display the per protocol matrix

```{r pp}
writeLines('Number of patients per arm in modified intention to treat analysis')
table(PP$Trt, Include_mITT = PP$Include_mITT)

writeLines('Number of swabs per protocol per treatment')
table(PP$Trt, PP_swabs = PP$Per_protocol_sample)
```




```{r data_summaries}
IDs = unique(platcov_dat$ID)
platcov_dat$Rand_date = as.POSIXct(platcov_dat$Rand_date)

writeLines(sprintf('We have %s PCR datapoints on %s patients from %s sites between %s and %s',
                   nrow(platcov_dat),
                   length(IDs),
                   length(unique(platcov_dat$Site)),
                   min(platcov_dat$Rand_date),
                   max(platcov_dat$Rand_date)))

# baseline samples: taken within 6 hours of randomisation
baseline_ind = platcov_dat$Timepoint_ID==0

bvl = aggregate(log10_viral_load ~ ID + Site, 
                platcov_dat[baseline_ind, ], median)
nrow(bvl) == length(IDs)

writeLines(sprintf('In the %s patients, the geometric mean baseline (defined as samples taken within 6 hours of randomisation) viral load was %s copies per mL (95 percentile interval from %s to %s; range from %s to %s)',
                   length(IDs),
                   round(10^mean(bvl$log10_viral_load)),
                   round(10^(mean(bvl$log10_viral_load)-1.96*sd(bvl$log10_viral_load))),
                   round(10^(mean(bvl$log10_viral_load)+1.96*sd(bvl$log10_viral_load))),
                   round(min(10^bvl$log10_viral_load)),
                   round(max(10^bvl$log10_viral_load))))
```


```{r variants_time}
par(las=1, cex.lab=1.5, cex.axis=1.5, family='serif')
var_cols = RColorBrewer::brewer.pal(n = length(unique(platcov_dat$Variant)),
                                    name = 'Set2')
ind_dup = !duplicated(platcov_dat$ID)
plot(platcov_dat$Rand_date[ind_dup], 1:length(IDs),
     col= adjustcolor(var_cols[as.numeric(as.factor(platcov_dat$Variant[ind_dup]))],.5),
     pch = 15-platcov_dat$Variant_Imputed[ind_dup]*15+1,cex=1.5,
     xlab='Enrollment date', ylab='Patient ID',panel.first=grid())
legend('topleft',col = var_cols,
       legend = levels(as.factor(platcov_dat$Variant[ind_dup])),
       pch=16,inset = 0.03,cex=2)
legend('left',col=1, pch=c(16,1),cex=2,
       legend = c('Genotyped','Imputed'),inset = 0.03)

table(Arm=platcov_dat$Trt[ind_dup], 
      Variant=platcov_dat$Variant[ind_dup], 
      Imputed=platcov_dat$Variant_Imputed[ind_dup])

table(platcov_dat$Site[ind_dup], platcov_dat$Variant_Imputed[ind_dup])
```


Summary table

```{r table1}
print(unique(platcov_dat$Trt))
trts = c("Regeneron",
         "Ivermectin",
         "Remdesivir",
         "Favipiravir",
         "Fluoxetine",
         'Molnupiravir',
         'Nirmatrelvir + Ritonavir',
         'Nitazoxanide',
         'Evusheld')
platcov_dat$Trt_code = 
  factor(platcov_dat$Trt,
         levels=c("No study drug", trts))

xx=make_baseline_table(input_data = platcov_dat[platcov_dat$ID %in% PP$ID[PP$Include_mITT],])
knitr::kable(xx, caption = sprintf('Summary of patient characteristics included in the current interim analysis (n= %s). Age: median (range); baseline viral load (log10 copies per mL: mean (range)); vaccinated: %% with any number of doses; number of vaccine doses: median (range); antibody data are from rapid tests done at screening (+ is presence of IgM or IgG band).', sum(PP$Include_mITT)))
```

## Summary data plot

This includes all patients in the mITT population with the per protocol swabs.

```{r trt_data_plot, fig.width=10, fig.height=7}
par(las=1, cex.lab=1.5, cex.axis=1.5, family='serif', mfrow=c(1,2))
ind_plot = platcov_dat$ID %in% PP$ID[PP$Include_mITT]
plot_baseline_data(platcov_dat[ind_plot,])
mtext(text = 'a',side = 3,adj = 0,cex=1.5,line = 1.5)

trt_cols = brewer.pal(n = length(unique(platcov_dat$Trt)),
                      name = 'Paired')
names(trt_cols) = c('No study drug', # all treatments
                    "Favipiravir",
                    'Fluoxetine',
                    'Nirmatrelvir + Ritonavir',
                    'Molnupiravir',
                    'Nitazoxanide',
                    'Regeneron',
                    'Remdesivir',
                    'Ivermectin',
                    'Evusheld')

plot_serial_data(xx = platcov_dat[ind_plot,],
                 trt_cols = trt_cols)
mtext(text = 'b',side = 3,adj = 0,cex=1.5,line = 1.5)
```


### Saliva versus swab

```{r saliva_PLATCOV, fig.height=5}
# IDs_saliva = unique(platcov_dat$ID[platcov_dat$Swab_ID == 'Saliva'])
# par(mfrow=c(2,3),las=1,family='serif',cex.axis=1.5, cex.lab=1.5)
# for(id in IDs_saliva){
#   ind = platcov_dat$ID==id
#   plot(platcov_dat$Time[ind], platcov_dat$log10_viral_load[ind],
#        col=as.numeric(platcov_dat$Swab_ID[ind]%in% c('Saliva','SAL'))+1,
#        panel.first=grid(),main=id,cex=1.5,pch=16,
#        xlab='Days from randomisation', ylab = 'Log10 viral load (copies per ml)')
# }
```



# Initial analysis

We fit a linear model to all data to check for:

* Major temporal trends
* Sanity check treatment effects


```{r brm_naive}
# scode <- make_stancode(log10_viral_load | cens(censor) ~ Time + (1+Time | ID),
#                        data = platcov_dat, family = 'student',
#                        prior = c(prior(lkj(2), class = "cor"),
#                                  prior(normal(5, 5), class = Intercept),
#                                  prior(constant(7), class = 'nu')))
platcov_dat$Trt = 
  factor(platcov_dat$Trt, 
         levels = c("No study drug",
                    "Ivermectin",
                    "Regeneron",
                    "Remdesivir",
                    "Favipiravir",
                    "Nitazoxanide",
                    "Fluoxetine",
                    "Molnupiravir",
                    "Nirmatrelvir + Ritonavir",
                    "Evusheld"))
platcov_dat$Epoch=as.factor(platcov_dat$Epoch)
# 
# ind_fit = platcov_dat$Timepoint_ID<14 & platcov_dat$Site=='br003'
# sum(ind_fit)
# if(RUN_MODELS){
#   mod_naive = brm(log10_viral_load | cens(censor) ~ Time*Trt-Trt +
#                     Time*Variant+Time*Epoch+(1+Time|ID),
#                   data = platcov_dat[ind_fit, ],
#                   prior = c(prior(lkj(2), class = "cor"),
#                             prior(normal(5, 5), class = Intercept),
#                             prior(constant(7), class = 'nu')),
#                   family = 'student', cores = 4, chains = 4)
#   save(mod_naive, file = 'Rout/naive_mod.RData')
# } else {
#   load( file = 'Rout/naive_mod.RData')
# }
# plot(mod_naive, pars=c('Trt', 'Variant','Epoch'), N=4)
# thetas = ranef(mod_naive)
# slopes_randef = thetas$ID[,1,'Time']
# 
# slopes_dat = merge(data.frame(ID=names(slopes_randef), slope_re = slopes_randef),
#                    platcov_dat[!duplicated(platcov_dat$ID),], by = 'ID')
# slopes_dat$Variant = 
#   factor(slopes_dat$Variant,
#          levels = c('Delta','BA.2','BA.1','BA.4','BA.5'))
# summary(lm(slope_re ~ Trt_code+Variant+Epoch, data = slopes_dat))
# 
# summary(lm(slope_re ~ Variant, data = slopes_dat[slopes_dat$Trt=='No study drug',]))
# 
# ind_not_BA4 = slopes_dat$Variant != 'BA.4'
# 
# mod0 = lm(slope_re ~ Trt+Variant, 
#           data = slopes_dat[ind_not_BA4& 
#                               slopes_dat$Trt%in% c('No study drug','Regeneron'),])
# mod1 = lm(slope_re ~ Trt*Variant, 
#           data = slopes_dat[ind_not_BA4& 
#                               slopes_dat$Trt%in% c('No study drug','Regeneron'),])
# summary(mod1)
# lmtest::lrtest(mod0, mod1)
```


```{r rlm}
IDs = PP$ID[PP$Include_mITT]
naive_slopes = array(dim=length(IDs))
names(naive_slopes)=IDs
for(id in IDs){
  ind = platcov_dat$ID==id & 
    platcov_dat$Time<8  
  cens_val = platcov_dat$log10_cens_vl[ind][1]
  if(all(platcov_dat$log10_viral_load[ind]>platcov_dat$log10_cens_vl[ind])){
    mod = lm(log10_viral_load ~ Time + CT_RNaseP, 
             data = platcov_dat[ind,],)
  } else {
    mod = censReg(log10_viral_load ~ Time + CT_RNaseP, 
                  left = cens_val,data = platcov_dat[ind,])
  }
  naive_slopes[id] = coef(mod)['Time']
}

naive_fit = merge(data.frame(ID=names(naive_slopes),slope=naive_slopes),
                  platcov_dat[!duplicated(platcov_dat$ID), ])
par(las=3, mar=c(12,5,2,2))
boxplot(slope~Trt, data = naive_fit,xlab='')
naive_fit$Trt=as.character(naive_fit$Trt)
boxplot(slope~Variant+Trt, 
        data = naive_fit[naive_fit$Trt%in%c('No study drug',
                                            'Regeneron'),],xlab='')
```


# Model fitting

The primary analysis consists of fitting a sequence of Bayesian hierarchical models to the mITT population (viral load data up until day 7, only patients who did not leave or switch before day 3).

There are a series of model fits, which are combinations of three models:

-   Model 1: vanilla student-t regression with left censoring at 0 and with individual random effects for slope an intercept;
-   Model 2: Linear model with additional RNaseP adjustment (this needs special treatment as we need to subtract the effect of RNaseP for the linear predictions);
-   Model 3: Non-linear model (up and then down) with RNaseP adjustment.

two sets of priors:

-   Weakly informative priors
-   Quasi flat prior (multiply SDs by 10)

and two sets of covariate adjustments:

-   basic covariate adjustment (slope and intercept):
-   Variant (main Greek lineages as defined by WHO)
-   Site
-   Full covariate adjustment (slope and intercept), adding in:
-   Vaccination (number of doses)
-   Age (standardized to have mean=0 and sd=1)
-   Time since symptom onset (days, between 0 and 4)
-   Serology rapid test (+/-)
-   Serum antibody: not yet decided which antibody will be measured - expected June 2022

## Specify priors

get priors

```{r priors}
source('priors.R')
# print(all_priors)
```

## Prepare models

We make the stan data sets.

```{r}
# All analysis datasets - data_prep.R does the construction of datasets
analysis_datasets = c(interim="interim_dat.csv",
                      favi = "Favipiravir_analysis.csv" ,
                      fluox = "Fluoxetine_analysis.csv" ,
                      molnpvr = "Molnupiravir_analysis.csv",
                      paxlovid = "Paxlovid_analysis.csv",
                      pax_mol = "Paxlovid_Molnupiravir_analysis.csv",
                      nitzox = "Nitazoxanide_analysis.csv")
trts_lists = list(c('No study drug', # all treatments
                    "Favipiravir",
                    'Fluoxetine',
                    'Nirmatrelvir + Ritonavir',
                    'Molnupiravir',
                    'Nitazoxanide',
                    'Regeneron',
                    'Remdesivir',
                    'Ivermectin',
                    'Evusheld'),
                  c('No study drug', "Favipiravir"),
                  c('No study drug', "Fluoxetine"),
                  c('No study drug', 'Molnupiravir'),
                  c('No study drug', 'Nirmatrelvir + Ritonavir'),
                  c('No study drug','Molnupiravir', 'Nirmatrelvir + Ritonavir'),
                  c('No study drug', 'Nitazoxanide'))
names(trts_lists) = names(analysis_datasets)
print(trts_lists)
platcov_dat_list = list()
ind_fitting_list = list()
for(i in 1:length(analysis_datasets)){
  analysis_name = names(analysis_datasets)[i]
  platcov_dat_list[[i]] = read.csv(file = analysis_datasets[i])
  
  writeLines(sprintf('Time since symtpm onset not yet entered for %s patients', sum(is.na(platcov_dat_list[[i]]$Symptom_onset[!duplicated(platcov_dat_list[[i]]$ID)]))))
  platcov_dat_list[[i]]$Symptom_onset[is.na(platcov_dat_list[[i]]$Symptom_onset)]=1
  
  
  writeLines(sprintf('In %s there are %s PCR datapoints and %s patients',
                     analysis_datasets[i], 
                     nrow(platcov_dat_list[[i]]),
                     length(unique(platcov_dat_list[[i]]$ID))))
  ## re-arrange so that censored values come last
  platcov_dat_list[[i]] = dplyr::arrange(platcov_dat_list[[i]], log10_viral_load==log10_cens_vl)
  ind_change_time = is.na(platcov_dat_list[[i]]$Time) | platcov_dat_list[[i]]$Time <0
  platcov_dat_list[[i]]$Time[ind_change_time] = platcov_dat_list[[i]]$Timepoint_ID[ind_change_time]
  # Datapoints that will be used in the analysis
  ind_fitting_list[[i]] = platcov_dat_list[[i]]$Time < Dmax & 
    platcov_dat_list[[i]]$Per_protocol_sample==1 &
    platcov_dat_list[[i]]$ID %in% PP$ID[PP$Include_mITT]
  
  platcov_dat_list[[i]]$Trt = factor(platcov_dat_list[[i]]$Trt, levels = trts_lists[[analysis_name]])
  writeLines(sprintf('Analysis dataset contains %s patients and %s datapoints',
                     length(unique(platcov_dat_list[[i]]$ID[ind_fitting_list[[i]]])),
                     sum(ind_fitting_list[[i]])))
}
names(platcov_dat_list) = names(analysis_datasets)
names(ind_fitting_list) = names(analysis_datasets)

covs_base = c('Variant','Site')
covs_full=c(covs_base, 'Age_scaled')
stan_inputs = list()

for(nn in names(analysis_datasets)){
  i = which(nn==names(analysis_datasets))
  stan_inputs[[i]] = 
    make_stan_inputs(input_data_fit = platcov_dat_list[[nn]][ind_fitting_list[[nn]], ],
                     int_covs_base = covs_base,
                     int_covs_full = covs_full,
                     slope_covs_base = covs_base,
                     slope_covs_full = covs_full,
                     trt_frmla = formula('~ Trt'),
                     Dmax = Dmax)
}
names(stan_inputs) = names(analysis_datasets)

# Add Epoch variable to full interim analysis
platcov_dat_list[['interim']]$Epoch = as.factor(platcov_dat_list[['interim']]$Epoch)
stan_inputs[['interim']] = 
  make_stan_inputs(input_data_fit = platcov_dat_list[['interim']][ind_fitting_list[['interim']], ],
                   int_covs_base = c('Site','Variant','Epoch'),
                   int_covs_full = c('Site','Variant','Epoch',
                                     'Age_scaled','Symptom_onset'),
                   slope_covs_base = c('Site','Variant','Epoch'),
                   slope_covs_full = c('Site','Variant','Epoch',
                                       'Age_scaled','Symptom_onset'),
                   trt_frmla = formula('~ Trt'),
                   Dmax = Dmax)
```


## Raw data on viral clearance

```{r viral_clearance_data}
par(las=1, cex.lab=1.5, cex.axis=1.5, family='serif', mfrow=c(1,1))
for(dd in names(analysis_datasets)[-1]){
  plot_serial_data(xx = platcov_dat_list[[dd]],
                   trt_cols = trt_cols)
}
```



## Setup model runs

List all parameters settings for the model fitting:

```{r setup_models}
all_mods = list.files('Stan_models',full.names = TRUE,pattern = '*stan')

model_settings = expand.grid(mod = all_mods[2:3],
                             prior = 1,
                             cov_matrices = 1,
                             dataset = analysis_datasets)
# expand.grid(mod = all_mods[2],
#       prior = 1,
#       cov_matrices = 2,
#       dataset = analysis_datasets))

model_settings$Niter = 2000
model_settings$Nwarmup = 1000
model_settings$Nthin = 4
model_settings$Nchain = 4

writeLines(sprintf('We are running all models with %s chains and %s samples for each chain, discarding %s for burn-in and thining every %s, thus giving a total of %s posterior samples per model.',
                   unique(model_settings$Nchain),
                   unique(model_settings$Niter),
                   unique(model_settings$Nwarmup),
                   unique(model_settings$Nthin), 
                   unique(model_settings$Nchain*(model_settings$Niter-model_settings$Nwarmup)/model_settings$Nthin)))


save(model_settings, 
     platcov_dat_list,
     analysis_datasets,
     ind_fitting_list,
     stan_inputs, 
     all_priors,
     file = 'Rout/model_run_setup.RData')
```

All model fitting is run on a cluster using run_models.R (bmrc.sh provides bash commands)

Load model fits

```{r}
ff = list.files('Rout/')
ff = ff[grep(pattern = 'model_fits_',x = ff)]
if(!length(ff)==nrow(model_settings)) stop('not all outputs are ready for all model settings')
ff = paste0('Rout/',ff)
```

# Results
## Model fits:

```{r summary_plots, fig.width=10, fig.height=7}
trt_effect_thetas = effect_ests = list()
for(i in 1:nrow(model_settings)){
  load(paste0('Rout/model_fits_',i,'.RData'))
  trt_effect_thetas[[i]] = extract(out,pars='trt_effect')$trt_effect
}

for(i in 1:nrow(model_settings)){
  dd = model_settings$dataset[i]
  col_names_trt = colnames(stan_inputs[[dd]]$Trt_matrix)[-1]
  col_names_trt=gsub(pattern = 'Trt',replacement = '',x = col_names_trt)
  col_names_trt=gsub(pattern = 'Variant',replacement = '',x = col_names_trt)
  col_names_trt=gsub(pattern = ':',replacement = ' ',x = col_names_trt)
  col_names_trt=gsub(pattern = '+',replacement = '+\n',x = col_names_trt,fixed = T)
  if(!is.null(dim(trt_effect_thetas[[i]]))){
    effect_ests[[i]] = t(apply(trt_effect_thetas[[i]], 2, 
                               quantile, probs=c(0.025,.1,.5,.9,.975)))
    rownames(effect_ests[[i]])=col_names_trt
    colnames(effect_ests[[i]])[3]='mean'
  }
}

par(las=1, mar=c(5, 12, 2,2), family='serif')
ind = which(model_settings$dataset==analysis_datasets[1] & model_settings$prior==1)
plot_effect_estimates(effect_ests = effect_ests,
                      plot_models = ind,
                      study_threshold = study_threshold,
                      my_pch = 15+as.numeric(model_settings$mod[ind]==all_mods[3]),
                      mod_cols = brewer.pal(n = 8,name = 'Dark2')[1:length(ind)])

for(i in 2:length(analysis_datasets)){
  par(las=1, mar=c(5, 12, 2,2), mfrow=c(1,2), family='serif')
  
  ind = which(model_settings$dataset==analysis_datasets[i] & model_settings$prior==1)
  plot_effect_estimates(effect_ests = effect_ests,
                        plot_models = ind,
                        study_threshold = study_threshold,
                        my_pch = 15+as.numeric(model_settings$mod[ind]==all_mods[3]),
                        mod_cols = brewer.pal(n = 8,name = 'Dark2')[1:length(ind)])
  par(mar=c(5,5,2,2))
  plot_serial_data(xx = platcov_dat_list[[names(analysis_datasets)[i]]],
                   trt_cols = trt_cols)
}
```


## Estimated treatment effects under all models

```{r treatment_effects}
for(i in 1:length(model_settings$dataset)){
  probs_sup = round(100*apply(trt_effect_thetas[[i]],2,function(x) mean(x > log(study_threshold))),1)
  names(probs_sup) = trts_lists[[model_settings$dataset[i]]][-1]
  trt_outputs = round(100* exp(apply(trt_effect_thetas[[i]], 2, quantile, probs=c(0.025,.5,.975))) - 100, 1)
  
  writeLines(sprintf('*********** RESULTS FOR %s **************', model_settings$dataset[i]))
  for(j in 1:length(probs_sup)){
    writeLines(sprintf('Under %s, probability %s increases rate of clearance >%s%% is %s%%',
                       model_settings$mod[i],
                       names(probs_sup)[j],
                       100*(study_threshold-1),
                       probs_sup[j]))
    writeLines(sprintf('For %s the increase in rate of clearance is %s (95%% CI %s to %s)',
                       names(probs_sup)[j],
                       trt_outputs[2,j],
                       trt_outputs[1,j],
                       trt_outputs[3,j]))
    
  }
}
```

Molnupiravir vs paxlovid 

```{r molnupiravir_v_paxlovid, fig.height=6, fig.width=9}
load(paste0('Rout/model_fits_',which(model_settings$dataset=='Paxlovid_Molnupiravir_analysis.csv')[1],'.RData'))
thetas_trt = extract(out, pars='trt_effect')$trt_effect
writeLines(sprintf('Molnupiravir has a lower effect than paxlovid with probability %s',
                   mean(thetas_trt[,1]<thetas_trt[,2])))

par(mfrow=c(1,2), las=1, mar=c(5, 10, 2,2))
ind = which(model_settings$dataset=='Paxlovid_Molnupiravir_analysis.csv' &
              model_settings$prior==1)
plot_effect_estimates(effect_ests = effect_ests,
                      plot_models = ind,
                      study_threshold = study_threshold,
                      my_pch = 15+as.numeric(model_settings$mod[ind]==all_mods[3]),
                      mod_cols = brewer.pal(n = 8,name = 'Dark2')[1:length(ind)])
par( mar=c(5, 4, 2,2))
hist(100*(exp(thetas_trt[,1])-exp(thetas_trt[,2])),
     main='',  xlab = 'Difference in treatment effect', 
     ylab='', yaxt='n')
abline(v=0, lwd=3, col='red', lty=2)

load(paste0('Rout/model_fits_',which(model_settings$dataset=='Paxlovid_Molnupiravir_analysis.csv')[2],'.RData'))
thetas_trt = extract(out, pars='trt_effect')$trt_effect
writeLines(sprintf('Molnupiravir has a lower effect than paxlovid with probability %s',
                   mean(thetas_trt[,1]<thetas_trt[,2])))
```




```{r favirapivir_weight_halflife}
load(paste0('Rout/model_fits_',which(model_settings$dataset=='Favipiravir_analysis.csv')[1],'.RData'))
slopes_fvp = colMeans(extract(out,pars='slope')$slope)
t12_output = data.frame(
  ID_stan = stan_inputs$favi$analysis_data_stan$id[stan_inputs$favi$analysis_data_stan$ind_start],
  t12 = 24*log10(.5)/slopes_fvp)
t12_output = merge(t12_output, stan_inputs$favi$ID_map, by='ID_stan')
t12_output = merge(platcov_dat_list$favi[!duplicated(platcov_dat_list$favi$ID), ],
                   t12_output, by.x = 'ID', by.y = 'ID_key')


summary(lm(t12 ~ Weight, data = t12_output[t12_output$Trt=='Favipiravir',]))
par(las=1)
plot(t12_output$Weight, t12_output$t12,
     col=as.numeric(t12_output$Trt=='Favipiravir')+1,
     xlab='Patient weight', panel.first=grid(), ylab=expression(t[1/2]))
```




## Individual plots

Individual plots colored by model

```{r individ_fits}
set.seed(4648865);
mod_cols = rep(sample(brewer.pal(n = 9, name = 'Set1')),2)
ID_map = stan_inputs$interim$ID_map
ind_not_dup = which(!duplicated(platcov_dat_list$interim$ID))
ID_map = merge(platcov_dat_list$interim[ind_not_dup, ], ID_map, by.x = 'ID',by.y = 'ID_key')

out_all = list()
load('Rout/model_fits_1.RData')
out_all[[1]] = out

load('Rout/model_fits_2.RData')
out_all[[2]] = out


ID_map = dplyr::arrange(ID_map, Site, Trt)

par(las=1, mfrow=c(5,5), mar=c(3,3,3,1),bty='n',
    cex.lab=1.5, cex.axis=1.5,family='serif')
plot_individ_data(mod_out = out_all,
                  models_plot = 1:2,
                  K_plots = 25,
                  mod_cols = mod_cols[1:2],
                  ID_map = ID_map,
                  analysis_data_stan = stan_inputs$interim$analysis_data_stan)

```

# Sensitivity analysis

## BRMS

```{r brms}
# platcov_dat$cen1 = ifelse(platcov_dat$log10_cens_vl==platcov_dat$log10_viral_load,
#                           'left', 'none')
# mod_exp = brms::brm(log10_viral_load | cens(cen1) ~ Time*Trt_code_subgroup - Trt_code_subgroup + (1+Time | Site + Variant), family='student', data = platcov_dat[ind_fitting, ], cores=4, chains = 4)
```


## lme4cens

```{r lem4cens_analysis}
# # Datapoints that will be used in the analysis
# ind_fitting = platcov_dat$Time < Dmax & 
#   platcov_dat$Per_protocol_sample==1 &
#   platcov_dat$ID %in% PP$ID[PP$Include_mITT]

# writeLines(sprintf('Total number of datapoints up until day %s is %s',
#                    Dmax, sum(ind_fitting)))
# yL=platcov_dat$log10_viral_load[ind_fitting]
# yL_cens=as.numeric(!platcov_dat$log10_viral_load>platcov_dat$log10_cens_vl)[ind_fitting]
# X = model.matrix(~Time*Trt_code-Trt_code, data = platcov_dat[ind_fitting, ])
# Z = matrix(rep(1, length(yL)), ncol=1)
# cluster = as.numeric(as.factor(platcov_dat$ID[ind_fitting]))
# fit = lmec(yL=yL,cens=yL_cens, X=X, Z=Z, cluster=cluster)

# platcov_dat$cens = 1
# platcov_dat$cens[platcov_dat$log10_cens_vl==platcov_dat$log10_viral_load]=2
# mod_cens=lmercens(Surv(log10_viral_load,cens) ~ 1+Time*Trt_code-Trt_code + (1|ID),
#                   data = platcov_dat[ind_fitting, ], REML = F)
# 
#                   # control = lme4::lmerControl(optimizer = "optimx",
#                   #                             optCtrl = list(method = "BOBYQA")))
# summary(mod_cens)

# 
# mod_lme4 = lme4::lmer(log10_viral_load ~ 1+Time*Trt_code-Trt_code + (1+Time|ID), 
#                       data = platcov_dat[ind_fitting, ])
# summary(mod_lme4)$coefficients
# fixef(mod_lme4)
```
