library(tidyverse)
library(stringr)
library(purrr)
library(readxl)

Reformat_Mutations <- function(nextclade_file, naming_file) {
  #import the nextclade data
  nextclade <- read_tsv(nextclade_file)
  
  #take the GISAID names from Excel andmerge to add PLATCOV naming scheme
  naming_cols <- c("PatientID","GISAID_sub","name","GISAID_ID","Genbank_sub","Genbank_ID")
  naming <- read_excel(naming_file, col_names=naming_cols)
  nextclade <- nextclade %>% separate_wider_delim(seqName, "/", names=c("hcov","loc","name", "year"))
  joined <- left_join(nextclade, naming)
  
  #select amino acid changes only and sort names
  aachanges <- joined %>% select(c(PatientID, aaSubstitutions))
  aachanges <- aachanges %>% separate_wider_delim(PatientID, "_", names=c("Samplename","Location", "Timepoint"))
  
  data_split <- aachanges %>%
    mutate(items = str_split(aaSubstitutions, ",")) %>%
    unnest(cols = items, keep_empty = TRUE)
  
  # create binary matrix with item names as column headers
  binary_matrix <- data_split %>%
    mutate(value = TRUE) %>%
    spread(key = items, value = value, fill = FALSE)
  #binary_matrix <- data_split %>%
  #  mutate(value = TRUE) %>%
  #  pivot_wider(names_from = items, values_from = value, values_fill=FALSE)
  
  return(binary_matrix)

}

Query_Mutations<- function(aa_matrix, querystring) {
  #extract the mutation columns which match the query string
  #if there is a wildcard and we accept any mutation at the final position
  if (endsWith(querystring, "*")) {
    #knock off the final * for querying
    querystring <- str_sub(querystring, 1, -2)
    #select all matching columns
    aa_matches <- aa_matrix %>% select(c(Samplename, Location, Timepoint,starts_with(querystring)))
    #merge column if any of the columns are true
    aa_matches <- aa_matches %>%
      mutate(any_true = rowSums(select(aa_matches, -c(Samplename, Location, Timepoint))) > 0)
    #drop the columns for each individual mutation
    aa_matches <- aa_matches %>% select(c('Samplename', 'Location','Timepoint','any_true'))
    #rename the column using the queryname
    colnames(aa_matches)[colnames(aa_matches) == "any_true"] =paste0(querystring,"_any") 
  }
  else {
    aa_matches <- aa_matrix %>% select(c(Samplename, Location, Timepoint,starts_with(querystring)))  
    
  }
  
  return(aa_matches)
}

#use the TSV file generated by standard Nextclade run
#eg,  nextclade run --input-dataset nextclade_datasets/ --output-all nextclade all_platcov_20230511.fasta
nextclade_file <- "nextclade/nextclade.tsv"
naming_file <- "Accession numbers_GISAID_GENBANK_PLATCOV_10-05-2023.xlsx"
aa_matrix <- Reformat_Mutations(nextclade_file, naming_file)
#query string should be in the format of gene:mutation, with * as a wildcard matching all mutations
querystring <- "S:N460*"
mutations <- Query_Mutations(aa_matrix, querystring)

querystring <- "ORF1a:T170I"
mutations <- Query_Mutations(aa_matrix, querystring)

#evusheld:
#any change in 486 
#OR any change in 346 or 444
evusheld_346 <- Query_Mutations(aa_matrix,"S:R346*")
evusheld_444 <- Query_Mutations(aa_matrix,"S:K444*")
evusheld_486 <- Query_Mutations(aa_matrix,"S:F486*")

#S3 Tixagevimab/cilgavimab
#The tixagevimab/cilgavimab resistant hypothesised subgroup is defined by: 
# (i) any amino-acid change from wild-type in the 486 residue of the spike protein AND (ii) an amino-acid change in the 346 residue OR the 444 residue.

evusheld_muts <- full_join(evusheld_346, evusheld_444)
evusheld_muts <- full_join(evusheld_muts, evusheld_486)

evusheld_muts <- evusheld_muts %>% mutate(
  evusheld_test = case_when(
    (`S:F486_any` == TRUE) & (`S:R346_any` == TRUE |`S:K444_any` == TRUE) ~ TRUE,
    TRUE ~ FALSE
  )
)  
write_csv(evusheld_muts, "evusheld_test.csv")